velox/functions/prestosql/ArrayAllMatch.cpp

```diff
@@ -0,0 +1,130 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/ScopedVarSetter.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+#include "velox/vector/FunctionVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+class AllMatchFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    exec::LocalDecodedVector arrayDecoder(context, *args[0], rows);
+    auto& decodedArray = *arrayDecoder.get();
+
+    auto flatArray = flattenArray(rows, args[0], decodedArray);
+    auto offsets = flatArray->rawOffsets();
+    auto sizes = flatArray->rawSizes();
+
+    std::vector<VectorPtr> lambdaArgs = {flatArray->elements()};
+    auto numElements = flatArray->elements()->size();
+
+    SelectivityVector finalSelection;
+    if (!context.isFinalSelection()) {
+      finalSelection = toElementRows<ArrayVector>(
+          numElements, *context.finalSelection(), flatArray.get());
+    }
+
+    VectorPtr matchBits;
+    auto elementToTopLevelRows = getElementToTopLevelRows(
+        numElements, rows, flatArray.get(), context.pool());
+
+    // Loop over lambda functions and apply these to elements of the base array,
+    // in most cases there will be only one function and the loop will run once.
+    context.ensureWritable(rows, BOOLEAN(), result);
+    auto flatResult = result->asFlatVector<bool>();
+    exec::LocalDecodedVector bitsDecoder(context);
+    auto it = args[1]->asUnchecked<FunctionVector>()->iterator(&rows);
+
+    // Turns off the ThrowOnError flag and temporarily resets errors in the
+    // context to nullptr, so we only handle errors in this function itself.
+    ScopedVarSetter throwOnError(context.mutableThrowOnError(), false);
+    ScopedVarSetter<exec::EvalCtx::ErrorVectorPtr> errorsSetter(
+        context.errorsPtr(), nullptr);
+    while (auto entry = it.next()) {
+      auto elementRows =
+          toElementRows<ArrayVector>(numElements, *entry.rows, flatArray.get());
+      auto wrapCapture = toWrapCapture<ArrayVector>(
+          numElements, entry.callable, *entry.rows, flatArray);
+      entry.callable->apply(
+          elementRows,
+          finalSelection,
+          wrapCapture,
+          &context,
+          lambdaArgs,
+          elementToTopLevelRows,
+          &matchBits);
+
+      bitsDecoder.get()->decode(*matchBits, elementRows);
+      entry.rows->applyToSelected([&](vector_size_t row) {
+        auto errors = context.errors();
+        if (errors && row < errors->size() && !errors->isNullAt(row)) {
```

Comment:
This logic doesn't look right. Errors for individual array elements should be suppressed only if the outcome can be decided by some other array element, e.g. if there is another element that returns 'false' for the predicate.

Reply:
Got it, I've fixed it and added the error suppression and rethrow test cases for it.
```SQL
select all_match(array[1, null, 0], x -> ((10 / x) > 2)); -- throw
select all_match(array[1, 0], x -> ((10 / x) > 2)); -- throw
select all_match(array[2, 5, 0], x -> ((10 / x) > 2)); -- false
select all_match(array[5, null, 0], x -> ((10 / x) > 2)); -- false
```